virttest/qemu_vm.py

```diff
@@ -1687,6 +1687,25 @@ def sort_key(dev):
                 vcpu_dev = devices.vcpu_device_define_by_params(params, vcpu)
                 devices.insert(vcpu_dev)
 
+        # Add device intel-iommu, it must be added before any virtio devices
+        if params.get('virtio_dev_iommu_platform') == 'on':
```

Comment:
The namespace of each virtio device should have its own `virtio_dev_iommu_platform` and it's only used to add `iommu_platform` option on virtio devices, so I suggest to introduce a new parameter to define `intel-iommu`.
Another issue is `intel-iommu` only available on x86, this change will break the initialization of other architectures.

Reply:
BTW, VGA device can be virtio-gpu-pci (aarch64 is using it), so `intel-iommu` must be added earlier than VGA.
https://github.com/avocado-framework/avocado-vt/blob/f425ead735daa1ccca408d8f62bdeb65c4b1f688/virttest/qemu_vm.py#L1487-L1489
https://github.com/avocado-framework/avocado-vt/blob/f425ead735daa1ccca408d8f62bdeb65c4b1f688/virttest/qemu_vm.py#L1109-L1120

Reply:
> The namespace of each virtio device should have its own `virtio_dev_iommu_platform` and it's only used to add `iommu_platform` option on virtio devices, so I suggest to introduce a new parameter to define `intel-iommu`.

@vivianQizhu I have same idea with @PaulYuuu , 'iommu_platform' is a virtio device option, use prefix `virtio_dev`.
Add "intel-iommu" device but without set 'iommu_platform' for virtio device is possible. It is better to introduce a new parameter to define `intel-iommu` device.

Reply:
> The namespace of each virtio device should have its own `virtio_dev_iommu_platform` and it's only used to add `iommu_platform` option on virtio devices, so I suggest to introduce a new parameter to define `intel-iommu`.
> Another issue is `intel-iommu` only available on x86, this change will break the initialization of other architectures.

@PaulYuuu Thanks for your review. Actually all test cases that specify this `virtio_dev_iommu_platform` variable they specify `only q35` together, so it wont affect other architectures, but yes, it should be restrict here. However, even I name a new parameter, it have same issue that it should be restrict to intel, so I would like to add a condition check here, instead of introduce a new parameter, what do you think?

Reply:
> BTW, VGA device can be virtio-gpu-pci (aarch64 is using it), so `intel-iommu` must be added earlier than VGA.

Yes, good catch.

Reply:
> > The namespace of each virtio device should have its own `virtio_dev_iommu_platform` and it's only used to add `iommu_platform` option on virtio devices, so I suggest to introduce a new parameter to define `intel-iommu`.
> 
> @vivianQizhu I have same idea with @PaulYuuu , 'iommu_platform' is a virtio device option, use prefix `virtio_dev`.
> Add "intel-iommu" device but without set 'iommu_platform' for virtio device is possible. It is better to introduce a new parameter to define `intel-iommu` device.

Well, okay, let me introduce a new parameter then. Thanks. cc @PaulYuuu

Reply:
> > The namespace of each virtio device should have its own `virtio_dev_iommu_platform` and it's only used to add `iommu_platform` option on virtio devices, so I suggest to introduce a new parameter to define `intel-iommu`.
> > Another issue is `intel-iommu` only available on x86, this change will break the initialization of other architectures.
> 
> @PaulYuuu Thanks for your review. Actually all test cases that specify this `virtio_dev_iommu_platform` variable they specify `only q35` together, so it wont affect other architectures, but yes, it should be restrict here. However, even I name a new parameter, it have same issue that it should be restrict to intel, so I would like to add a condition check here, instead of introduce a new parameter, what do you think?

@vivianQizhu Maybe we can check if the device in the qemu device category has `intel-iommu`?
ppc:
```
Uncategorized devices:
name "nvdimm", desc "DIMM memory module"
name "pc-dimm", desc "DIMM memory module"
name "spapr-tpm-proxy"
```

x86:
```
Uncategorized devices:
name "edu", bus PCI
name "intel-iommu", bus System
name "nvdimm", desc "DIMM memory module"
name "pc-dimm", desc "DIMM memory module"
```

And `virtio_dev_iommu_platform` should work on other archs.
```
# /usr/libexec/qemu-kvm -M pseries -device virtio-net-pci,iommu_platform=on,disable-legacy=on
VNC server running on ::1:5900
```

Reply:
@PaulYuuu yes, good idea. Will update later, thanks.